### PR TITLE
perf(471): Add duration metrics for wasm abi calls

### LINKED ISSUE #471: Record duration metric for wasm abi calls
This should include the time spent serializing rows from the database and copying into wasm memory.
Label each recording with transaction type, database address, reducer name, and syscall.

### PR DESCRIPTION
**Pull request by joshua-spacetime**
Closes #471.

# Description of Changes

# API and ABI

 - [ ] This is a breaking change to the module ABI
 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

# Expected complexity level and risk

*How complicated do you think these changes are? Grade on a scale from 1 to 5,
where 1 is a trivial change, and 5 is a deep-reaching and complex change.*

*This complexity rating applies not only to the complexity apparent in the diff,
but also to its interactions with existing and future code.*

*If you answered more than a 2, explain what is complex about the PR,
and what other components it interacts with in potentially concerning ways.*

### CODE REVIEW

File: crates/core/src/host/mod.rs [hunk 1]
```diff
@@ -166,3 +167,12 @@ impl EnergyMonitor for NullEnergyMonitor {
     ) {
     }
 }
+
+#[derive(Debug, Display)]
+pub enum SysCall {
```
Comment (kulakowski) [inline crates/core/src/host/mod.rs:172]: I'd love to reconcile this with https://github.com/clockworklabs/SpacetimeDB/blob/c566dbae3dc759b3f712cd0fc4a5f172d32306ed/crates/core/src/host/wasm_common/instrumentation.rs#L84

File: crates/core/src/util/typed_prometheus.rs [hunk 2]
```diff
@@ -121,6 +121,7 @@ impl_prometheusvalue_string!(
     Identity,
     Address,
     TransactionType,
+    SysCall,
```
Comment (kulakowski) [inline crates/core/src/util/typed_prometheus.rs:124]: We've waffled a lot on calling this syscalls vs abicalls vs... what. And even in this PR you say SysCall here but wasm_abi_call elsewhere.

Let's pick one.

**Review by kulakowski — COMMENTED** [review]
The rest lgtm tho

File: crates/core/src/host/wasmer/wasm_instance_env.rs [hunk 3]
```diff
@@ -436,15 +460,36 @@ impl WasmInstanceEnv {
...
-            let count = caller
-                .data()
-                .instance_env
-                .delete_by_col_eq(&ctx, table_id.into(), col_id.into(), &value)?;
-            Ok(count.get())
-        })
+        let ctx = caller.data().reducer_context();
+        let txn_type = &ctx.txn_type();
+        let db = &ctx.database();
+        let reducer = &ctx.reducer_name().unwrap_or_default().to_owned();
```
Comment (kulakowski) [inline crates/core/src/host/wasmer/wasm_instance_env.rs:466]: The reducer name copies are a bit unfortunate. Are they necessary? Is this what the static metric stuff is about?

https://crates.io/crates/prometheus-static-metric

This is definitely addressable via a followup if so.

**Review by kulakowski — APPROVED** [review]

File: crates/core/src/host/wasmer/wasm_instance_env.rs (hunk 3 above)
Comment (kulakowski) [reply crates/core/src/host/wasmer/wasm_instance_env.rs:466]: Ah right, I was thinking about just the abi call names. Yeah, sounds good.

File: crates/core/src/util/typed_prometheus.rs (hunk 2 above)
Comment (cloutiertyler) [reply crates/core/src/util/typed_prometheus.rs]: I would argue it's really a syscall where SpacetimeDB is the OS. Happy to resolve the bikeshedding in a separate PR though.

**Pull request merged**